interactive_engine/executor/ir/graph_proxy/src/apis/graph/element/path.rs

```diff
@@ -26,46 +25,85 @@ use ir_common::generated::algebra::path_expand::ResultOpt;
 use ir_common::generated::results as result_pb;
 use ir_common::{LabelId, NameOrId};
 use pegasus::codec::{Decode, Encode, ReadExt, WriteExt};
-use pegasus_common::downcast::*;
+use pegasus_common::downcast::Any;
+use pegasus_common::downcast::AsAny;
 use pegasus_common::impl_as_any;
 
-use crate::apis::{Element, GraphElement, PropertyValue, Vertex, ID};
+use crate::apis::{Edge, Element, GraphElement, PropertyValue, Vertex, ID};
+
+#[derive(Clone, Debug, Hash, PartialEq, PartialOrd)]
+pub enum VertexOrEdge {
+    V(Vertex),
+    E(Edge),
+}
+
+impl From<Vertex> for VertexOrEdge {
+    fn from(v: Vertex) -> Self {
+        Self::V(v)
+    }
+}
+
+impl From<Edge> for VertexOrEdge {
+    fn from(e: Edge) -> Self {
+        Self::E(e)
+    }
+}
+
+impl VertexOrEdge {
+    pub fn as_vertex(&self) -> Option<&Vertex> {
+        match self {
+            VertexOrEdge::V(v) => Some(v),
+            _ => None,
+        }
+    }
+
+    pub fn as_edge(&self) -> Option<&Edge> {
+        match self {
+            VertexOrEdge::E(e) => Some(e),
+            _ => None,
+        }
+    }
+}
 
 #[derive(Clone, Debug)]
 pub enum GraphPath {
-    AllV(Vec<Vertex>),
-    SimpleAllV(Vec<Vertex>),
-    EndV((Vertex, usize)),
-    SimpleEndV((Vertex, Vec<ID>)),
+    /// Arbitrary path, which may contain both vertices and edges, or only vertices.
+    AllPath(Vec<VertexOrEdge>),
+    /// Simple path, which may contains both vertices and edges, or only vertices.
+    SimpleAllPath(Vec<VertexOrEdge>),
+    /// Arbitrary path with only end vertices preserved, which may contain both vertices and edges, or only vertices.
+    EndV((VertexOrEdge, usize)),
+    /// Simple path with only end vertex preserved, which may contains both vertices and edges, or only vertices.
+    SimpleEndV((VertexOrEdge, Vec<ID>)),
 }
 
-impl_as_any!(GraphPath);
-
 impl GraphPath {
-    pub fn new(entry: Vertex, path_opt: PathOpt, result_opt: ResultOpt) -> Self {
+    pub fn new<E: Into<VertexOrEdge>>(entry: E, path_opt: PathOpt, result_opt: ResultOpt) -> Self {
         match result_opt {
             ResultOpt::EndV => match path_opt {
```

Comment:
This ResultOpt and PathOpt are both from pb, includes the whole path, like: `pb::xx::xx::ResultOpt`. In addition, I know `ResultOpt::AllVe` is generated, but can we avoid generating this name (that is inconsistent with our naming convention).

Reply:
Done